src/main/java/fi/otavanopisto/kuntaapi/server/rest/ServicesApiImpl.java

```diff
@@ -0,0 +1,510 @@
+    }
+    
+    ServiceLocationChannelId serviceLocationChannelId = toServiceLocationChannelId(serviceLocationChannelIdParam);
+    if (serviceLocationChannelId == null) {
+      return createBadRequest(String.format(INVALID_SERVICE_LOCATION_CHANNEL_ID, serviceIdParam));
+    }
+    
+    for (ServiceChannelProvider serviceChannelProvider : getServiceChannelProviders()) {
+      ServiceLocationChannel serviceLocationChannel = serviceChannelProvider.findServiceLocationChannel(serviceId, serviceLocationChannelId);
+      if (serviceLocationChannel != null) {
+        return Response.ok(serviceLocationChannel)
+            .build();
+      }
+    }
+    
+    return Response
+      .status(Status.NOT_FOUND)
+      .entity(NOT_FOUND)
+      .build();
+  }
+
+  @Override
+  public Response findServiceWebPageChannel(String serviceIdParam, String webPageChannelIdParam) {
+    ServiceId serviceId = toServiceId(serviceIdParam);
+    if (serviceId == null) {
+      return createBadRequest(String.format(INVALID_SERVICE_ID, serviceIdParam));
+    }
+    
+    WebPageChannelId webPageChannelId = toWebPageChannelId(webPageChannelIdParam);
+    if (webPageChannelId == null) {
+      return createBadRequest(String.format(INVALID_WEBPAGE_CHANNEL_ID, serviceIdParam));
+    }
+    
+    for (ServiceChannelProvider serviceChannelProvider : getServiceChannelProviders()) {
+      WebPageChannel webPageChannel = serviceChannelProvider.findWebPageChannelChannel(serviceId, webPageChannelId);
+      if (webPageChannel != null) {
+        return Response.ok(webPageChannel)
+            .build();
+      }
+    }
+    
+    return Response
+      .status(Status.NOT_FOUND)
+      .entity(NOT_FOUND)
+      .build();
+  }
+
+  @Override
+  public Response listServiceElectronicChannels(String serviceIdParam, Long firstResult, Long maxResults) {
+    ServiceId serviceId = toServiceId(serviceIdParam);
+    if (serviceId == null) {
+      return createBadRequest(String.format(INVALID_SERVICE_ID, serviceIdParam));
+    }
+    
+    if (firstResult != null && firstResult < 0) {
+      return createBadRequest(FIRST_RESULT_MUST_BY_A_POSITIVE_INTEGER);
+    }
+    
+    if (maxResults != null && maxResults < 0) {
+      return createBadRequest(MAX_RESULTS_MUST_BY_A_POSITIVE_INTEGER);
+    }
+    
+    List<ElectronicChannel> result = new ArrayList<>();
+    
+    for (ServiceChannelProvider serviceChannelProvider : getServiceChannelProviders()) {
+      result.addAll(serviceChannelProvider.listElectronicChannels(serviceId));
+    }
+    
+    int resultCount = result.size();
+    int firstIndex = firstResult == null ? 0 : Math.min(firstResult.intValue(), resultCount);
+    int toIndex = maxResults == null ? resultCount : Math.min(firstIndex + maxResults.intValue(), resultCount);
+    
+    return Response.ok(result.subList(firstIndex, toIndex))
+      .build();
+  }
+
+  @Override
+  public Response listServicePhoneChannels(String serviceIdParam, Long firstResult, Long maxResults) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) The Cyclomatic Complexity of this method "listServicePhoneChannels" is 12 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)